File: Calling C from Lua/bag.lua
local ffi = require("ffi")
local C = ffi.C

-- Extension for shared libraries
local libext = "so"
if ffi.os == "OSX" then
  libext = "dylib"
end

local libbag = ffi.load("./libbag." .. libext)

ffi.cdef[[
  /* From our library */

  typedef struct Bag Bag;
  typedef int value_type;
  typedef int size_type;

  Bag* new_bag();
  void delete_bag(Bag* b);

  void insert(Bag* b,value_type element);
  bool erase_one(Bag* b,value_type element);
  size_type erase_all(Bag* b,value_type element);
  size_type size(Bag* b);
  size_type count(Bag* b,value_type element);
  void print(Bag* b);
  
  /* From the C library */
  void free(void*);
]]


local function  bag_insert(self,target)
	libbag.insert(self.super,target)
	
end

local function bag_erase_one(self,target)
  return libbag.erase_one(self.super, target)
end

local function bag_erase_all(self, target)
  return libbag.erase_all(self.super, target)
end

local function bag_size(self)
  return libbag.size(self.super)
end

local function bag_count(self, target)
  return libbag.count(self.super, target)
  end

local function bag_print(self)
	libbag.print(self.super);
end

local bag_methods={
bag_insert=bag_insert,
bag_erase_one=bag_erase_one,
bag_erase_all=bag_erase_all,
bag_size=bag_size,
bag_count=bag_count,
bag_print=bag_print
}


local new=function(opts)
	local self={super=libbag.new_bag()}
	ffi.gc(self.super,libbag.delete_bag)
	return setmetatable(self,{ __index = bag_methods });
end 


return {new=new};

File: Calling C from Lua/bag.cpp
#include "bag.h"
#include <cassert>

	void Bag::insert(const Bag::value_type& j)
	{
		assert(used<CAPACITY);
		data[used] = j;
		++used;
	}

	bool Bag::erase_one(const Bag::value_type& j)
	{
		size_type indx = -1;
		for(size_type i = 0; i < used; i++)
		{
			if(data[i] == j)
			{
				indx = i;
				break;
			}
		}
		if(indx == -1)
			return false;
		data[indx] = data[used-1];
		--used;
		return true;
	}

	size_t Bag::erase_all(const Bag::value_type& j)
	{
		size_t total = 0;
		while (erase_one(j))
			total++;
		return total;
	}

	Bag::size_type Bag::count(const Bag::value_type& j) const
	{

		Bag::size_type total = 0;
		for (int i = 0; i < used; i++)
		{
			if (j == data[i])
			{
				total++;
			}
		}
		return total;
	}

	void Bag::operator+=(const Bag& addend)
	{
		assert(used + addend.used < CAPACITY);
		for (int i = 0; i < addend.used; i++)
		{
			//insert(addend.data[i]);
			data[used] = addend.data[i];
			used++;
		}
	}

	void Bag::print() const
	{
		for (int i = 0; i < used; i++)
		{
			std::cout << data[i] << " ";
		}
		std::cout << std::endl;

	}
              
	size_t Bag::size() const
	{
		return used;
	}
